### PR TITLE
feat: get cdd claims from storage

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/base/Context.ts [hunk 1]
```diff
@@ -607,58 +611,97 @@ export class Context {
...
+          claim_type: claimTypeToMeshClaimType(ClaimType.CustomerDueDiligence, this),
+        },
+        paginationOpts: {
+          size: size ?? DEFAULT_GQL_PAGE_SIZE,
+        },
+      });
+
+      const data: ClaimData[] = [];
+      entries.forEach(([key, identityClaim]) => {
+        const claim1stKey = key.args[0] as Claim1stKey;
```
Comment (monitz87) [inline src/base/Context.ts:688]: You should destructure this

File: src/base/Context.ts [hunk 2]
```diff
@@ -607,58 +611,97 @@ export class Context {
...
+          // eslint-disable-next-line @typescript-eslint/camelcase
+          claim_type: claimTypeToMeshClaimType(ClaimType.CustomerDueDiligence, this),
+        },
+        paginationOpts: {
+          size: size ?? DEFAULT_GQL_PAGE_SIZE,
+        },
+      });
+
+      const data: ClaimData[] = [];
+      entries.forEach(([key, identityClaim]) => {
```
Comment (monitz87) [inline src/base/Context.ts:687]: You should destructure identityClaim

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/base/Context.ts [hunk 3]
```diff
@@ -606,59 +653,75 @@ export class Context {
...
-    });
+      const next = calculateNextKey(count, size, start);
 
-    const next = calculateNextKey(count, size, start);
+      return {
+        data,
+        next,
+        count,
+      };
+    } else {
```
Comment (monitz87) [inline src/base/Context.ts:713]: This else is not necessary, there is a return at the end of the previous block

File: src/base/Context.ts [hunk 4]
```diff
@@ -606,59 +653,75 @@ export class Context {
...
+      }
 
-    return {
-      data,
-      next,
-      count,
-    };
+      const identityClaimsEntries = await this.identityClaimsEntries({ targets, claimTypes });
+
+      return identityClaimsEntries;
```
Comment (monitz87) [inline src/base/Context.ts:723]:
```suggestion
      return this.identityClaimsEntries({ targets, claimTypes });
```

File: src/base/__tests__/Context.ts [hunk 5]
```diff
@@ -1174,13 +1182,13 @@ describe('Context class', () => {
...
 
-      let result = await context.issuedClaims({
+      let result = (await context.issuedClaims({
         targets: [targetDid],
         trustedClaimIssuers: [targetDid],
         claimTypes: [ClaimType.Accredited],
         includeExpired: true,
         size: 1,
-      });
+      })) as ResultSet<ClaimData>;
```
Comment (monitz87) [inline src/base/__tests__/Context.ts:1191]: These assertions shouldn't be necessary once the function is returning a single type

File: src/base/__tests__/Context.ts [hunk 6]
```diff
@@ -1200,12 +1208,108 @@ describe('Context class', () => {
         }
       );
 
-      result = await context.issuedClaims();
+      result = (await context.issuedClaims()) as ResultSet<ClaimData>;
```
Comment (monitz87) [inline src/base/__tests__/Context.ts:1211]: These assertions shouldn't be necessary once the function is returning a single type

File: src/base/__tests__/Context.ts [hunk 7]
```diff
@@ -1200,12 +1208,108 @@ describe('Context class', () => {
...
 
       expect(result.data).toEqual(fakeClaims);
       expect(result.count).toEqual(25);
       expect(result.next).toBeNull();
     });
+
+    test('should return a list of claims', async () => {
```
Comment (monitz87) [inline src/base/__tests__/Context.ts:1218]: Merge these two tests

File: src/base/__tests__/Context.ts [hunk 8]
```diff
@@ -1200,12 +1208,108 @@ describe('Context class', () => {
...
+
+      const result = (await context.issuedClaims({
+        targets: [targetDid],
+        claimTypes: [ClaimType.CustomerDueDiligence],
+      })) as ClaimData[];
+
+      expect(result).toEqual(fakeClaims);
+    });
+
+    test('should throw if the middleware query fails and targets or claimTypes are not seted', async () => {
```
Comment (monitz87) [inline src/base/__tests__/Context.ts:1296]:
```suggestion
    test('should throw if the middleware is not available and targets or claimTypes are not set', async () => {
```

File: src/base/__tests__/Context.ts [hunk 9]
```diff
@@ -1266,7 +1370,7 @@ describe('Context class', () => {
       });
     });
 
-    test('should return a list of claims', async () => {
+    test('should return a result set of claims', async () => {
```
Comment (monitz87) [inline src/base/__tests__/Context.ts:1373]: Isn't this test exactly the same as the first one?

File: src/base/Context.ts [hunk 10]
```diff
@@ -585,6 +590,48 @@ export class Context {
     return assembleResult(didRecords);
   }
 
+  /**
+   * @hidden
+   */
+  private async identityClaimsEntries(args: {
```
Comment (monitz87) [inline src/base/Context.ts:596]: getIdentityClaimsFromChain

File: src/base/Context.ts [hunk 11]
```diff
@@ -585,6 +590,48 @@ export class Context {
...
+      polymeshApi: {
+        query: { identity },
+      },
+    } = this;
+
+    const { targets, claimTypes } = args;
+    const data: ClaimData[] = [];
+
+    await P.each(targets, async rawTarget => {
+      await P.each(claimTypes, async rawClaimType => {
```
Comment (monitz87) [inline src/base/Context.ts:610]: Just `claimType`

Comment (monitz87) [inline src/base/Context.ts:610]: Double each is too slow. There should be a way of doing this using P.map

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/base/Context.ts [hunk 12]
```diff
@@ -585,6 +591,60 @@ export class Context {
...
+              target: new Identity({ did: identityIdToString(target) }, this),
+              issuer: new Identity({ did: identityIdToString(claimissuer) }, this),
+              issuedAt: momentToDate(issuanceDate),
+              expiry,
+              claim: meshClaimToClaim(claim),
+            });
+          }
+        }
+      );
+    });
```
Comment (monitz87) [inline src/base/Context.ts:643]: This approach may lead to inconsistent ordering. You should have the map function return the claim data if `(!includeExpired && (expiry === null || expiry > new Date())) || includeExpired` is true, and undefined if not, and then filter out the undefined values afterwards with lodash `compact`

File: src/base/Context.ts [hunk 13]
```diff
@@ -609,55 +669,79 @@ export class Context {
...
+    }
+
+    if (!targets || !claimTypes) {
+      throw new PolymeshError({
+        code: ErrorCode.FatalError,
+        message: 'Cannot perform this action without an active middleware connection',
+      });
+    }
+
+    const identityClaimsFromChain = await this.getIdentityClaimsFromChain({
```
Comment (monitz87) [inline src/base/Context.ts:735]: You're not filtering the results by claim issuers

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/base/Context.ts (hunk 13 above)
Comment (monitz87) [reply src/base/Context.ts:735]: The `getIdentityClaimsFromChain` function should do the filtering

File: src/base/Context.ts [hunk 14]
```diff
@@ -618,31 +623,40 @@ export class Context {
...
-        }
+        )
       );
+
+      return compactData;
     });
 
-    return data;
+    return flattenDeep(claimData).filter(({ issuer }) =>
+      trustedClaimIssuers ? trustedClaimIssuers.includes(issuer.did) : true
```
Comment (monitz87) [inline src/base/Context.ts:658]: `trustedClaimIssuers` can also contain Identities

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/base/Context.ts (hunk 14 above)
Comment (shuffledex) [reply src/base/Context.ts:658]: line 626:

```
    const trustedClaimIssuers = rawTrustedClaimIssuers?.map(trustedClaimIssuer =>
      signerToString(trustedClaimIssuer)
    );
```

Comment (monitz87) [reply src/base/Context.ts:657]: Missed that. Call the original variable `trustedClaimIssuers` and this one `claimIssuerDids`

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/base/Context.ts [hunk 15]
```diff
@@ -629,32 +629,31 @@ export class Context {
...
+            });
+          } else {
+            data.push(undefined);
           }
-        )
+        }
       );
-
-      return compactData;
+      return compact(data);
```
Comment (monitz87) [inline src/base/Context.ts]: Since you're using `forEach`, you no longer need the `undefined` values nor `compact`

**Review by monitz87 — CHANGES_REQUESTED** [review]

**Review by monitz87 — APPROVED** [review]

**Pull request merged**